FEAT #1: Add member join function and test methods

B2SIC opened the pull request:
- #1

## 추가 내역
* 회원가입을 위한 Spring 기본 구성 요소
  * Member Controller
    * API를 통한 접근 방식을 채택
  * Member Domain
    * userId
    * password
    * phoneNumber
  * Member Find DTO
    * 회원 검색에 사용하기 위한 DTO
  * Member Repository
    * JPA Entity Mapping 방식을 사용
  * Member Service
    * 회원가입
    * 중복 회원 가입 방지
    * 전체 회원 검색
    * 회원 아이디로 회원 검색
    * 회원 휴대폰 번호로 회원 검색
* 회원 아이디 또는 휴대폰 번호로 회원을 검색할 수 있는 기능 
* 회원 가입 및 검색 기능 관련 테스트 코드
  * 회원 가입 성공 여부
  * 아이디가 중복인 회원 가입 실패 여부
  * 회원 아이디 기반 회원 검색
  * 회원 휴대폰 번호 기반 회원 검색
  * 전체 회원 검색 (데이터 개수 비교를 통한 테스트)
* build.gradle dependencies에 관련 Method 정의
  * Getter & Setter Annotation 사용을 위한 Lombok
  * H2 Database
    ![image](https://user-images.githubusercontent.com/28584171/190202334-d77fbfd9-46b9-43bc-98d9-8487297d870c.png)

  * Validation 관련
  * Spring 데이터 JPA (JPA 포함)
* application.properties에 DB 관련 정보 및 JPA 설정
* Spring Config
  * 향후 Repository 변경 가능성에 대응하기 위함

B2SIC committed:
FEAT #1: Add member join function and test methods
Add dependencies to build.gradle
Add datasource and jpa properties to application.properties
Add member controller
Add member domain
Add member repository
Add member service
Add member DTO
Add spring config file
Add member repository test methods
Add member service test methods

namsic commented on src/main/resources/application.properties:
```diff
@@ -1 +1,6 @@
+spring.datasource.url=jdbc:h2:tcp://localhost/~/test
+spring.datasource.driver-class-name=org.h2.Driver
+spring.datasource.username=sa
```
이 부분은 나중에 localhost가 아닌 다른 내용으로 변경하게 될 것 같은데,
DB 관련 내용이 외부에 공개되어도 문제가 없을지

B2SIC replied:
로컬이 아닌 서버 환경의 DB를 적용하게 될 경우 DB 관련 내용은 외부에 공개되면 안됩니다.
이 경우 별도의 properties 파일을 만들어서 관리하는 방법을 고려 해야 할 것으로 보입니다.

namsic commented on src/test/java/harmony/dev/harmonyserver/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,68 @@
+    @Autowired MemberRepository memberRepository;
+
+    @Test
+    public void findTest(){
```
아래 항목에 대한 Test는 필요하지 않은지
- find 결과가 존재하지 않는 경우
- find 결과가 2개 이상인 경우

B2SIC replied:
테스트 케이스는 앞으로 계속해서 추가하면서 정교하게 작성할 예정입니다.
말씀하신 테스트 케이스도 추가해서 반영하도록 하겠습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/service/MemberService.java:
```diff
@@ -0,0 +1,42 @@
+        this.memberRepository = memberRepository;
+    }
+
+    public String join(Member member) {
```
단순 중복 여부 외에 입력값에 대한 검증이 필요하지는 않은지
- userId 또는 password에 특수문자가 포함되는 경우
- phoneNumber에 숫자 외의 값이 포함되는 경우

필요하다면, 이에 대한 테스트 코드도 추가되어야 함

B2SIC replied:
DB 관련 필드를 정의할 때 Validation에 관한 고민이 있었습니다.

말씀하신 부분을 포함해서 사용자의 아이디와 비밀번호는 몇 자리로 제한을 하고 어떤 문자까지 허용 해야 할지
이런 부분이 세부적으로 정해져야 하는데, 요구 사항이 별도로 존재하지 않기 때문에 고려 대상에서 잠시 제외했습니다.

DB에 userId와 password는 각각 20자 까지 허용하도록 했고
phoneNumber의 경우 표준 규격에 따라 15자리까지 허용하도록 했습니다.
userId의 경우 특수 문자를 포함하지 않도록 제약을 걸어야 하겠지만, password는 포함을 하는 것이 일반적입니다.

현재까지 관련 사항들을 정리해보면 아래와 같습니다.

* userId는 특수 문자는 허용하지 않으며 20자리 까지만 허용함.
  * 일반적으로 -나 _의 특수문자 정도는 허용하는데 전부 허용하지 않는 쪽으로 해야 할지?
  * 이외에 다른 제약은 필요하지 않은지? (userId가 숫자로 시작 하는 등)
* password는 특수문자를 허용하며 20자리 까지만 허용함.
* phoneNumber는 오직 숫자만 허용함.
  * -(하이픈) -> Filtering 또는 입력 폼 단에서 처리 후 DB에 숫자만 저장하도록
* userId, phoneNumber 모두 공백 비허용 -> password는 허용 해야 하는가?

사용자 정보 필드에 대한 피드백이 필요합니다.
관련 내용이 상세히 정해지면 이에 대한 테스트 코드를 함께 추가하도록 하겠습니다.

## 관련 테스트 코드 추가 대상
* userId 또는 password에 특수문자가 포함되는 경우
* userId 또는 password 또는 phoneNumber에 정해진 자리 수를 초과한 길이의 값이 들어오는 경우
* phoneNumber에 숫자 이외에 문자가 들어오는 경우

namsic replied:
> 말씀하신 부분을 포함해서 사용자의 아이디와 비밀번호는 몇 자리로 제한을 하고 어떤 문자까지 허용 해야 할지 이런 부분이 세부적으로 정해져야 하는데, 요구 사항이 별도로 존재하지 않기 때문에 고려 대상에서 잠시 제외했습니다.

이에 대해 질문한 이유는 SQL injection 가능성 때문입니다. 아직 기준이 명확하지 않으므로 validation 관련 내용은 별도 issue 생성하여 다루도록 하고 이번 PR에서는 고려하지 않겠습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/controller/MemberController.java:
```diff
@@ -0,0 +1,38 @@
+    @PostMapping("/member/find")
+    @ResponseBody
+    public Optional<Member> find(@Valid @RequestBody MemberFindDTO memberFindDTO){
+        if(memberFindDTO.getType().equals("user_id")){
+            return memberService.findByUserId(memberFindDTO.getKey());
+        }else if(memberFindDTO.getType().equals("phone_number")){
+            return memberService.findByPhoneNumber(memberFindDTO.getKey());
+        }else{
+            throw new IllegalStateException("잘못된 타입입니다.");
```
```suggestion
        String dtoType = memberFindDTO.getType();
        Optional<Member> result = switch(dtoType) {
            case "user_id" -> memberService.findByUserId(memberFindDTO.getKey());
            case "phone_number" -> memberService.findByPhoneNumber(memberFindDTO.getKey());
            default -> null;
        }
        if(result == null) {
            throw new IllegalStateException("잘못된 타입입니다.");
        }
        return result;
```
위와 같은 형태로 변경하는 것이 어떤지
- getType() 1회만 수행
- 마지막에 return. 필요한 경우 return 이전에 find 결과에 대한 추가 작업 수행 가능

또는 MemberRepository의 find가 하나로 통합된다면 join과 마찬가지로
`return memberService.find(...);` 형태로 변경하고,
위 로직은 함수 내부에서 수행하도록 하는 방법도 고려할 필요 있음

B2SIC replied:
말씀하신대로 `memberFindDTO.getType()` 을 한 번만 1회만 수행 하는 코드가 더 좋은 코드라고 생각합니다.
MemberController의 find는 제안해주신 것처럼 변경해서 반영하도록 하겠습니다.

memberService 또는 MemberRepository의 find 코드를 하나로 통합하는 방법에 대해서는 조금 회의적입니다.
사용자가 회원 가입을 하는 동작을 예로 들었을 때, 아이디 중복 검사를 하는 과정이 필요합니다.
이 때는 정확히 findByUserId 함수의 동작 만을 필요로 하기 때문에 이 둘의 동작을 분리 시키고 Controller단에서 필요에 따라 합치는 것이 코드의 가독성 측면에서 더 좋지 않을까 생각합니다.

이에 대한 생각은 어떤지 의견 주시면 감사하겠습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/repository/DefaultMemberRepository.java:
```diff
@@ -0,0 +1,44 @@
+        return member;
+    }
+
+    @Override
+    public Optional<Member> findByUserId(String userId) {
+        List<Member> result = em.createQuery("SELECT m from Member m WHERE m.userId = :userId", Member.class)
+                .setParameter("userId", userId)
+                .getResultList();
+        return result.stream().findAny();
+    }
+
+    @Override
+    public Optional<Member> findByPhoneNumber(String phoneNumber) {
+        List<Member> result = em.createQuery("SELECT m from Member m WHERE m.phoneNumber = :phoneNumber", Member.class)
+                .setParameter("phoneNumber", phoneNumber)
+                .getResultList();
+        return result.stream().findAny();
+    }
```
- 아래와 같은 형태로 통합하는 것보다 각각을 분리하는게 더 좋은 패턴인지
```java
    @Override
    public Optional<Member> find(String column, String value) {
        List<Member> result = em.createQuery(
            "SELECT m from Member m WHERE m." + column + " = :param", 
            Member.class)
        .setParameter("param", value)
        .getResultList();
        return result.stream().findAny();
    }
```
- `findAny()`를 통해 하나의 값을 return하고 있는데, 
조회 결과가 1개가 아닌 경우에 하나의 값만 사용해도 문제가 없을지 모르겠음

B2SIC replied:
* 아래와 같은 형태로 통합하는 것보다 각각을 분리하는게 더 좋은 패턴인지
  * 위에서 남겨주신 피드백과 같은 결의 논의라고 생각합니다.
제 생각은 함수 별로 기능은 분리 되는 것이 코드 가독성 측면에서 좋다 라는 생각입니다.
이와 관련해서 여러 사람들의 의견을 들어보고 싶습니다.

* `findAny()`를 통해 하나의 값을 return하고 있는데,
조회 결과가 1개가 아닌 경우에 하나의 값만 사용해도 문제가 없을지 모르겠음
  * find의 경우 현재 userId와 phoneNumber 두 가지를 찾아주는 기능을 하는데, 두 가지 모두 결국 DB 안에서는 Unique 해야한다고 생각했습니다. 그래서 결과 값의 개수는 0 또는 1의 경우만 생각하고 구현을 했는데 혹시 둘 이상의 결과가 return 될 경우도 고려하는 것이 맞을까요? 회원가입을 승인 하는 단계에서 예방이 가능할 것으로 보이지만 다른 좋은 방법이 있는지 찾아보겠습니다.

namsic replied:
현재 findByUserId와 findByPhoneNumber는 WHERE의 column을 제외하고 완전히 동일한 로직을 수행하고 있습니다.
이후에 Member에 다른 field가 추가되는 경우에도 유사한 코드가 중복으로 추가될 것 같은데,
외부에서 사용 시 함수가 구분되는 것이 유용하다면 아래와 같은 형태의 구현은 어떤지 검토 바랍니다.
```java
    private List<Member> findByValue(String column, String value) {
        return em.createQuery("SELECT m from Member m WHERE m." + column + " = :value", Member.class)
            .setParameter("value", value)
            .getResultList();
    }

    private Optional<Member> findUniqueByValue(String column, String value) {
        List<Member> findResult = findByValue(column, value);
        if(findResult.size() > 1) { /* exception */ }
        return findResult.stream().findAny();
    }

    @Override
    public Optional<Member> findByUserId(String userId) {
        return findUniqueByValue("userId", userId);
    }

    @Override
    public Optional<Member> findByPhoneNumber(String phoneNumber) {
        return findUniqueByValue("phoneNumber", phoneNumber);
    }
```

> 두 가지 모두 결국 DB 안에서는 Unique 해야한다고 생각했습니다.

DB 내부에서 unique한 값을 가지는 경우 현재 구현으로도 충분하지만 만약 이후에 DB가 변경되어 해당 값이 중복으로 존재하는 상황이 생기면 겉으로 드러나는 에러나 예외는 없으면서 기대하지 않은 동작을 수행하게 될 것 같은데, 이에 대해 고려할 필요는 없나요?

namsic commented on src/main/java/harmony/dev/harmonyserver/DTO/MemberFindDTO.java:
```diff
@@ -0,0 +1,15 @@
+import javax.validation.constraints.NotEmpty;
+
+@Getter
+@Setter
```
DTO는 데이터 교환을 위한 객체이므로 생성 이후에 값이 변경되는 상황이 없다.
setter 대신 DTO 생성 시점에 한 번만 초기화하는 방향이 적절할 것으로 보임

B2SIC replied:
동의합니다! 
제가 몰랐던 부분이었습니다.

MemberFindDTO의 경우 Post 방식으로 Mapping 되는 `/member/find`에서 사용되는데, POST의 경우 `@RequestBody` 를 통해 JSON 데이터가 넘어오고 이 데이터를 Java Object로 변환 해야 합니다. 이 때문에 Setter가 필요할 것이라고 생각했지만 이를 Spring에서 JSON 타입의 변환을 담당하는 `Jackson2HttpMessageConverter` 가 Converter 내에 존재하는 `ObjectMapper`로 값을 변환해주는 역할을 한다고 합니다. 따라서 Setter가 사용되지 않고 API 테스트를 통해 정상 작동 함을 확인했습니다.

반대로 Get 방식의 요청의 경우에는 `Query Parameter` 이기 때문에 `Jackson2HttpMessageConverter` 가 사용되지 않고 Spring 에서는 `WebDataBinder` 가 사용된다고 합니다. `WebDataBinder`는 별다른 설정이 없을 경우 Java Bean 방식으로 값을 할당하는데 Java Bean은 Setter를 사용해서 값을 할당 하기 때문에 Setter가 필요하다고 합니다.

이렇게 하나를 또 알아가는군요!

namsic commented on src/main/java/harmony/dev/harmonyserver/domain/Member.java:
```diff
@@ -0,0 +1,31 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+
+    @NotEmpty
+    @JsonProperty("user_id")
+    private String userId;
```
`@NotEmpty` 보다
`@Column(nullable = false, unique = true)` 등을 활용하는 것이 맞지 않나

B2SIC replied:
빈 문자열에 대응하기 위해 `@NotEmpty` 를 사용했습니다.
`@Column`의 옵션으로는 "" 와 같은 빈 문자열에 대응하지 못하는 것으로 알고 있습니다.

namsic replied:
> validation 관련 내용은 별도 issue 생성하여 다루도록 하고 이번 PR에서는 고려하지 않겠습니다.

이 때 함께 처리하도록 합시다.

namsic commented on src/main/java/harmony/dev/harmonyserver/domain/Member.java:
```diff
@@ -0,0 +1,31 @@
+import javax.validation.constraints.NotEmpty;
+
+@Getter
+@Setter
```
Entity에 setter가 필요한지 잘 모르겠음

B2SIC replied:
현재 구현 과정에서는 Member Domain에 `@Setter`는 사용이 안되고 있습니다.

하지만 테스트 코드에서 사용되고 있습니다.
`@Setter` 없이 Member 객체 내에 private 필드에 값을 주입하는 방법이 떠오르지 않는데
이 부분은 고민이 필요할 것 같습니다.

namsic replied:
- https://khdscor.tistory.com/15
- https://okky.kr/articles/807224

참고 바랍니다. private 값을 설정하기 위해 무조건 setter가 필요한 것은 아닙니다.

B2SIC replied:
Builder 패턴에 대한 공부가 필요 할 것으로 보입니다.
참고해서 이후에 관련 의견을 남기거나 반영 할 수 있도록 해보겠습니다.

namsic replied:
현재 Member의 attribute가 많지 않으므로 빌더패턴 없이 생성자에서 초기화해도 복잡하지 않을 것 같은데, 어떤가요?
```java
// setter
Member member1 = new Member();
member1.setUserId("testId1");
member1.setPassword("passwd");
member1.setPhoneNumber("01011111111");

// constructor
Member meber1 = new Member("testId1", "passwd", "01011111111");
```

B2SIC replied:
현재 상황이라면 직접 생성자를 만드는 것도 괜찮은 방법이라고 생각합니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/repository/MemberRepository.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface MemberRepository {
```
JpaRepository를 상속받는 것보다 필요한 부분만 구현하는게 더 유용한가?

B2SIC replied:
JpaRepository를 상속 받는다는게 무슨 말인지 잘 이해가 되지 않습니다.
좀 더 자세한 설명 부탁드립니다.

namsic replied:
```java
@Repository
public interface MemberRepository extends JpaRepository<Member, Long> {
    public List<Member> findByUserId(String userId);
    public List<Member> findByPhoneNumber(String phoneNumber);
}

```
참고
- https://ykh6242.tistory.com/104
- https://araikuma.tistory.com/329

B2SIC replied:
Spring 데이터 JPA를 적용하지 않은 것은 의도한 것입니다.
이는 JPA 숙련도와 관련이 있는데, 현재 단계에서 바로 Spring 데이터 JPA를 적용하는 것은 무리가 있다고 판단했습니다.
Spring 데이터 JPA의 경우 JPA에 대한 지식이 많이 없는 상태로 사용할 경우 오류에 대한 대응이 훨씬 어렵기 때문에
JPA에 익숙해지고 난 후 Spring 데이터 JPA를 차츰 적용해보는 방향으로 나갈 계획입니다.

namsic reviewed (commented):
1차 리뷰 완료
- 외부에서 API를 사용할 때 참고할 수 있는 자료가 있으면 좋을 듯

B2SIC committed:
STYLE #1: Change member find return type and code style
Change MemberController find method return type
Remove setter annotation in MemberFindDTO
Add conditions for find function result
Add validation code for phone number

B2SIC committed:
TEST #1: Change find and join test code
Change variable type in find test method
Add find test code for result value 0
Add validation test code for phone number duplication

B2SIC commented:
## 변경 사항
* find 함수 결과 타입
  ```
  {
      "message": "Success",
      "result": Object
  }
  ```
  * 일단 형식에 대한 정의를 해두고 향후 ResponseEntity와 같은 클래스를 사용해서 구현할 예정
* join 함수 결과 타입
  ```
  {
      "message": "OK"
  }
  ```
*  if -> switch ~ case 적용
* find 결과 조건 분기
  * 0개 : null
  * 1개 : member 객체 반환
  * 2개 이상 : 500 ERROR
* join 시 중복 검사 대상에 phoneNumber 추가
* find 결과 분기에 따른 테스트 코드 변수 타입 수정
* find 결과가 존재 하지 않을 때 테스트 코드 추가
* phoneNumber가 중복인 경우에 대한 테스트 코드 추가

## 그 외
* Entity에 Setter 사용은 Builder를 적용하거나 Setter를 테스트 코드에만 삽입하는 방법으로 대응 예정

namsic commented:
https://devlog-wjdrbs96.tistory.com/197
https://jeong-pro.tistory.com/200
api 응답 관련 구글링 자료입니다.
카카오 등 major api의 요청/응답 구조를 참고하는 것도 좋겠습니다.

> - find 함수 결과 타입
> ```json
> {
>     "message": "Success",
>     "result": Object
> }
> ```
> 일단 형식에 대한 정의를 해두고 향후 ResponseEntity와 같은 클래스를 사용해서 구현할 예정
> 
> - join 함수 결과 타입
> ```json
> {
>     "message": "OK"
> }
> ```
@regular1020 의견 바랍니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/controller/MemberController.java:
```diff
@@ -0,0 +1,59 @@
+        Map<String, Object> resultMap = new LinkedHashMap<>();
+
+        Optional<Member> result;
+        switch(memberFindDTOType) {
+            case "user_id":
+                result = memberService.findByUserId(memberFindDTO.getKey());
+                break;
+            case "phone_number":
+                result = memberService.findByPhoneNumber(memberFindDTO.getKey());
+                break;
+            default:
+                throw new IllegalStateException("잘못된 타입입니다.");
+        }
```
```suggestion
        result = switch(memberFindDTOType) {
            case "user_id" -> memberService.findByUserId(memberFindDTO.getKey());
            case "phone_number" -> memberService.findByPhoneNumber(memberFindDTO.getKey());
            default -> null;
        }
```
이전에 제안했던 enhanced switch expressions을 사용하지 않은 이유가 있나요?

B2SIC replied:
프로젝트가 Java 11을 기반으로 하고 있기 때문에 해당 문법은 사용이 불가 합니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/controller/MemberController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @PostMapping("/member/join")
+    @ResponseBody
+    public Map<String, String> join(@Valid @RequestBody Member member){
+        String result = memberService.join(member);
+        Map<String, String> resultMap = new LinkedHashMap<>();
+
+        if(result.equals("OK")){
+            resultMap.put("message", "OK");
+        }else{
+            resultMap.put("message", "Fail");
+        }
+        return resultMap;
+    }
```
`memberService.join(member)`에서는 특정 값이 중복인 경우
IllegalStateException과 메시지를 throw하고 있습니다.

- 예외 발생 시 어떤 값도 return하지 않는데 외부에서 중복된 값으로 join 요청을 보내면 응답이 어떻게 되나요?
  * Exception 발생했기 때문에 26번 line에서 실행 중단하고 `중복된 ..가 존재합니다` 메시지와 함께 에러 응답
  * result에 특정 값(empty string 등)이 들어가 `Fail` 메시지와 함께 200 정상 응답

전자의 경우 29-33의 조건문은 의미가 없는 코드이며,
후자의 경우 함수에서 발생한 예외는 현재 로직에서 굳이 필요하지 않은 동작입니다.

추가로 단순 성공/실패를 분기하려는 경우 `memberService.join()`의 return type은 boolean으로 변경하고,
Controller level에서 메시지를 입력하는 것이 좋겠습니다.

B2SIC replied:
* 이미 사용 중인 아이디로 회원가입을 시도하는 경우 HttpStatus Code 500과 메세지를 함께 리턴합니다.
  ![image](https://user-images.githubusercontent.com/28584171/191761006-33c61b67-79b4-40c0-904e-a6484b25ef46.png)

* 변경사항에 기재한 것처럼 리턴 값에 대해선 아직 논의된 형식이 없는 상태입니다.
  따라서 임시로 형식을 직접 설정 해둔 상태이며 이 부분은 향후 ResponseEntity를 사용해서 구현할 예정입니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/repository/DefaultMemberRepository.java:
```diff
@@ -0,0 +1,56 @@
+        return member;
+    }
+
+    @Override
+    public Optional<Member> findByUserId(String userId) {
+        List<Member> result = em.createQuery("SELECT m from Member m WHERE m.userId = :userId", Member.class)
+                .setParameter("userId", userId)
+                .getResultList();
+
+        if(result.size() <= 1){
+            return result.stream().findAny();
+        }else{
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "중복된 아이디가 존재합니다.");
+        }
+    }
+
+    @Override
+    public Optional<Member> findByPhoneNumber(String phoneNumber) {
+        List<Member> result = em.createQuery("SELECT m from Member m WHERE m.phoneNumber = :phoneNumber", Member.class)
+                .setParameter("phoneNumber", phoneNumber)
+                .getResultList();
+
+        if(result.size() <= 1){
+            return result.stream().findAny();
+        }else{
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "중복된 휴대폰 번호가 존재합니다.");
+        }
+    }
```
두 함수는 코드가 거의 동일한데, 함수 추출 or 추상화는 필요하지 않나요?

B2SIC replied:
함수 추출이라는 것이 이전 리뷰에서 말한 하나의 함수로 두 가지 동작을 같이 수행하도록 하는 방식인가요?

namsic replied:
findByPhoneNumber와 findByUserId는 구분하는 것이 더 적절하다 했으니
DefaultMemberRepository 내부에 공통 부분을 private method 형태로 분리하는 것입니다.

현재는 함수가 2개이니 큰 문제가 없지만, 이후 유사한 함수가 추가될 때 마다 아래와 같은 문제가 생기게 됩니다.
- 재사용성이 떨어지며, 동일한 코드를 반복해서 사용하는 것은 노동력이 많이 소모된다.
- 중복된 코드는 버그가 발견되거나 수정 사항이 발생할 때마다 동일한 수정 작업을 n번 반복해야 한다.
- 여러 위치에 흩어져 있기 때문에 로직에 대한 문서화 혹은 설명을 작성하기 어렵다.
- 중복된 상태로 그대로 두었다가 누군가 흩어진 중복 코드 중에서 한 곳만 고치는 경우,
원래의 의도와 달리 처음에는 같은 기능을 하던 코드들이 다른 방식으로 동작하게 되는 문제가 발생할 수 있다.

또한 책 `Clean Code`의 `3장. 함수`에서는
함수를 만들 때 첫 번째 규칙은 "**작게**"이며, 두 번째 규칙은 "**더 작게**"라고 소개하고 있습니다.

아래는 이전 리뷰에서 예시로 작성했던 코드입니다.
```java
    private List<Member> findByValue(String column, String value) {
        return em.createQuery("SELECT m from Member m WHERE m." + column + " = :value", Member.class)
            .setParameter("value", value)
            .getResultList();
    }

    private Optional<Member> findUniqueByValue(String column, String value) {
        List<Member> findResult = findByValue(column, value);
        if(findResult.size() > 1) { /* exception */ }
        return findResult.stream().findAny();
    }

    @Override
    public Optional<Member> findByUserId(String userId) {
        return findUniqueByValue("userId", userId);
    }

    @Override
    public Optional<Member> findByPhoneNumber(String phoneNumber) {
        return findUniqueByValue("phoneNumber", phoneNumber);
    }
```

namsic commented on src/test/java/harmony/dev/harmonyserver/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,75 @@
+
+        Optional<Member> userIdTest = memberRepository.findByUserId("testId2");
+        assertThat(userIdTest.isPresent()).isTrue();
+        assertThat(userIdTest.get()).isEqualTo(member2);
+        assertThat(userIdTest.get().getUserId()).isEqualTo(member2.getUserId());
```
- Member끼리 비교한 후에 userId를 다시 비교하는 이유는 무엇인가요?
- `get()` method 한 번만 수행하는 것이 좋겠습니다.

B2SIC replied:
별 다른 이유는 없습니다, 교차 테스트를 하다 보니 생긴 코드입니다.

namsic reviewed (commented):
2차 리뷰

regular1020 commented:
api 관련 의견 올립니다.

api관련해서 카카오api 레퍼런스, 네이버 파파고 api 레퍼런스 등을 살펴보고 몇 가지 검색해봤습니다.
https://developers.kakao.com/docs/latest/ko/reference/rest-api-reference
https://developers.naver.com/docs/papago/papago-detectlangs-api-reference.md

카카오와 네이버 둘 다 오류 발생 시 내부적으로 정한 별도의 오류 코드를 통해 구분합니다.
카카오 : 음수로 이뤄진 코드
네이버 : 영어 + 숫자로 이뤄진 코드

저희도 이와 같은 방식으로 에러 발생 시에는 코드를 구분하여 body 내부에 전달하는 방식이 좋을 것 같습니다.
현재 생각하고 있는 방식은 4자리 숫자로 만들어서 앞자리는 어떤 method인지, 뒷자리는 어떤 에러인지를 나타낼 수 있게 하는 방식입니다.
정상적으로 처리될 시에는 body 내부에 "result"와 같은 이름으로 object를 추가하여 전달하고, 메시지는 성공과 실패 모두 사용하며, 간단히 받는 개발자가 어떤 상황인지 알 수 있을 정도의 짧은 문장이면 충분할 것 같습니다.
또한 에러 코드는 정리한 뒤 따로 api관련 문서를 만들어서 포함하여 정리하는 것이 좋을 것 같습니다.

다음은 카카오 api의 레퍼런스 페이지를 바탕으로 제가 만든 회원 가입, 회원 조회 관련 문서 일부와 request, response 예시입니다.
# 문서

## 공통

### 공통 상태코드

|상태 코드|설명|
|------|---|
|200|요청을 성공적으로 수행|
|400|필수 파라미터 관련 오류 등 일반적인 오류로 인한 실패|
|500|서버 오류|

## 회원 가입

### 회원 가입 파라미터

|파라미터|타입|설명|필수|
|------|---|------|---|
|id|string|사용자 id|1|
|phone|string|사용자 전화번호|1|
|pw|string|사용자 비밀번호|1|

### 회원 가입 에러코드 10

|에러 코드|설명|
|------|---|
|01|중복 id로 회원가입|
|02|잘못된 전화번호로 회원가입|

### 회원가입 관련 method에 대한 response 예시
```
// request
{
    "id" : "abcd",
    "phone" : "01012345678",
    "pw" : "aofwnl12jol1029nla"
}

// 회원가입 성공
200 OK
{
    "message" : "OK"
}

// id 중복으로 인한 회원가입 실패
400 Bad Request
{
    "err-code" : 1001,
    "message" : "id already used"
}
```


## 회원 조회

|파라미터|타입|설명|필수|
|------|---|------|---|
|phone|string|검색하려는 사용자 전화번호|1|

### 회원 조회 에러코드 11
|에러 코드|설명|
|------|---|
|01|해당하는 회원이 존재하지 않음|

### 회원 검색 관련 method에 대한 response 예시
```
// request
{
    "phone" : "01012345678"
}

// 회원 검색 성공
200 OK
{
    "message" : "OK",
    "result" : {
        "id" : "abcd",
        "phone" : "010-1234-5678",
    }
}

// 해당 조건에 맞는 회원 정보가 존재하지 않아 회원 검색 실패
400 Bad Request
{
    "err-code" : 1101,
    "message" : "no such user"
}
```

일단 혼자 짧은 시간 공부하고 구상해본 api 형식이기에 완전하지 않을 수 있고, 서버 입장에서 생각해본건 아니기에 의견 주시면 감사하겠습니다.

namsic commented:
@regular1020 예시로 든 내용은 네이버의 api에서 메서드 부분을 숫자로 변경한 형태인 것 같습니다.
- 알파벳 두 자리의 경우 600개 이상의 조합이 가능한데, 우리 서비스에서는 100개면 표현하기에 충분한가요?
- 알파벳과 숫자를 사용하는 경우, 어디까지가 api code이고 어디부터 err code인지 명확하게 구별이 가능하고,
`AA11`, `AAA1`, `A111` 등 각 부분의 길이가 다른 코드가 섞여있는 경우에도 유연하게 처리할 수 있을 것으로 보입니다.
숫자형 코드 체계의 경우는 별도의 전처리 과정이 필요 없는 장점이 있을 것 같은데, 이런 부분도 고려되었나요?

@B2SIC @chanw12 위 내용 검토하고 의견 바랍니다.

regular1020 commented:
저는 서비스가 그렇게 크지 않기에 100개면 충분하다고 생각하였고 그렇다면 문자열이 섞여있는것보다 처리가 쉬울것으로 생각하여서 이러한 방식으로 제안했습니다.
또한 api코드와  err코드는 고정적으로 2자리씩 나눠서 사용한다고만 생각하였기에 AA11, AAA1, A111 등 각 부분의 길이를 다르게 사용하는 경우는 고려하지 않았습니다.
이러한 부분까지 고려한다면 api종류가 많지 않더라도 알파벳을 조합하여 사용하는 것이 더 좋을 수도 있을 것 같네요.
다른 분들의 의견도 종합적으로 들어보고 결정하는것이 좋을 것 같습니다.

B2SIC commented:
두 방식 모두 괜찮다고 생각합니다.
어차피 숫자만 있는 코드나 알파벳과 함께 조합된 코드나 결국 API 문서를 참고 하지 않으면
단 번에 알아보기는 어려울 것으로 예상됩니다.

다만 저의 경우는 에러 코드는 숫자로만 이루어진 것이 더 익숙하다고 할 수 있겠네요.
추가로 숫자나 문자에 따른 API 표현 개수는 신경 쓰지 않아도 될 문제라고 생각합니다.

namsic commented:
> 현재 생각하고 있는 방식은 4자리 숫자로 만들어서 앞자리는 어떤 method인지, 뒷자리는 어떤 에러인지를 나타낼 수 있게 하는 방식입니다.
정상적으로 처리될 시에는 body 내부에 "result"와 같은 이름으로 object를 추가하여 전달하고, 메시지는 성공과 실패 모두 사용하며, 간단히 받는 개발자가 어떤 상황인지 알 수 있을 정도의 짧은 문장이면 충분할 것 같습니다.

위 방식 사용하도록 합시다. 현재 api 두개 뿐이므로 예시의 코드를 그대로 사용해도 충분할 것 같습니다.

B2SIC committed:
STYLE #1: Change API return type and code style
Apply ResponseEntity to join and find function
Change domain fields annotation to clarify conditions
Add ResponseDTO to use ResponseEntity
Change the integration of similar code for find function
Change return type for status detail
Change application.properties -> application.yml for improved readability
Add Message domain

B2SIC committed:
TEST #1: Change object type and return type in test code

B2SIC commented:
## 변경사항

### API Response Type 수정
* ResponseEntity 적용
  * ResponseDTO 추가
  * 관련 코드 Return Type 및 테스트 코드 수정
  * Return 되는 Data에 일부 필드 Ignore 처리
    * ID (PK, not user_id)
    * Password
* join
  * 성공
  ```
  {
      "message": "OK"
  }
  ```
  
  * 실패
  ```
  {
      "message": "이미 사용 중인 아이디입니다.",
      "errorCode": "0101"
  }
  ```
  ```
  {
      "message": "이미 사용 중인 휴대폰 번호입니다.",
      "errorCode": "0102"
  }
  ```
* find
  * 성공
  ```
  {
      "message": "OK",
      "data":{
          "user_id": "test_id1",
          "phone_number": "01012345678"
      }
  }
  ``` 
  * 일치하는 데이터가 없음
    * Error로 분류하지 않은 이유는 find 동작 자체는 올바르게 수행되었기 때문.
  ```
  {
      "message": "OK",
      "data": null
  }
  ```
  * 잘못된 타입 입력
  ```
  {
      "message": "잘못된 타입입니다.",
      "errorCode": "0201"
  }
  ```
### 기타
* Member Domain 제약 추가 (길이, Blank 허용 여부 등)
* 함수 통합 예시 코드 반영
* application.properties -> application.yml 로 변경
* Message Domain 추가
  * 관련 내용은 현재 미구현 상태

namsic commented on src/main/java/harmony/dev/harmonyserver/repository/DefaultMemberRepository.java:
```diff
@@ -0,0 +1,67 @@
+    private Optional<Member> findUniqueByValue(String column, String value){
+        List<Member> findResult = findByValue(column, value);
+
+        if (findResult.size() > 1){
+            String message;
+            switch (column){
+                case "userId":
+                    message = "중복된 아이디가 존재합니다.";
+                    break;
+                case "phoneNumber":
+                    message = "중복된 휴대폰 번호가 존재합니다.";
+                    break;
+                default:
+                    message = "알 수 없는 오류가 발생했습니다.";
+            }
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, message);
+        }
```
위 부분은 findUniqueByValue() 내부에서 다루는 것보다 호출한 곳에서 처리하도록 넘기는 것이 어떤가요?
아래의 코드는 단순 로직을 표현하기 위함이며, 적용 시에는 java 및 spring 구조에 맞게 변경이 필요할 것 같습니다.

```java
    private Optional<Member> findUniqueByValue(String column, String value){
        List<Member> findResult = findByValue(column, value);
        if (findResult.size() > 1){
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "중복 값이 존재합니다.");
        }
        return findResult.stream().findAny();
    }
```
```java
// 호출하는 측에서 메시지 변경해 다시 던지기
    public Optional<Member> findByUserId(String userId){
        try {
            return memberRepository.findByUserId(userId);
        } catch(ResponseStatusException e) {
            throw new ResponseStatusException(e.getCode()/* 이런 getter 있는지는 모름 */, "userId duplicated.");
        }
    }
```

B2SIC replied:
음.. 저도 Exception이 발생하는 로직에 대해서는 고민 중입니다.
바로 밑에 주신 의견과 함께 처리되어야 할 부분으로 생각됩니다.
더 나은 방식이 있는지 함께 고려해보겠습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/service/MemberService.java:
```diff
@@ -0,0 +1,49 @@
+        return resultPair;
+    }
+
+    private Pair<Boolean, String> validateDuplicateMember(Member member) {
+        if (memberRepository.findByUserId(member.getUserId()).isPresent()){
+            return Pair.of(false, "userId");
+        }else if(memberRepository.findByPhoneNumber(member.getPhoneNumber()).isPresent()){
+            return Pair.of(false, "phoneNumber");
+        }else{
+            return Pair.of(true, "ok");
+        }
+    }
```
위의 함수는 아래와 같은 값을 return하는 것을 기대하고 있습니다.
- db에 같은 값이 있는 경우 `{false, <duplicated Key>}`
- db에 같은 값이 없는 경우 `{true, "ok"}`

db에 같은 값이 하나만 있는 경우 올바르게 동작하지만
2개 이상인 경우(현재는 불가능한 시나리오이기는 함)의 return은
`{false, <duplicated Key>}`가 아닌 것으로 보입니다.

이 부분은 고려하지 않아도 괜찮은가요?

B2SIC replied:
말씀하신 부분은 사전에 고려했던 상황인데, 아직 어떻게 처리할지 고민 중인 부분입니다.

같은 아이디 또는 휴대폰 번호가 2개 이상인 경우가 없는 것이 정상인 상황에서 결국 2개 이상이 나오는지 체크하기 위해서 Return Type을 전부 바꿔야 하는 것이 올바른가에 대한 고민이 있습니다.

DB에 같은 값이 존재하는 상황에 대해서 현재 예외 처리를 Repository 내에서 하고 있는데 단순히 지금처럼 Exception을 Throw 할 경우 기존에 논의된 API의 Return Type 처럼 나오지 않습니다. 이 때문에 Exception을 던지는 것이 아닌 ResponseEntity Type으로 만들어서 반환하는 방법, 그리고 발생된 Exception에 대한 ExceptionController를 두고 관리하는 방법 총 2가지를 생각했습니다.

우선 첫 번째 Repository 계층에서 ResponseEntity를 만들어서 반환하는 방법은 가능 할 것으로 보이긴 하지만 Response에 대한 응답을 Controller가 아닌 Repository에서 처리하는 것이 올바른가에 대한 고민입니다.
두 번째 ExceptionController를 두는 방법의 경우, Throw로 지정한 Exception에 대해선 모두 같은 처리가 되도록 구성이 되어야 합니다. 이것이 문제가 없는지 또, 이 때 Error 코드를 지정하는 일 만을 별도로 처리하는 것이 가능할 지에 대해서 찾아보고 가능할 경우 적용이 가능한 방법입니다.

해당 케이스 뿐만 아니라 예를 들어서, 들어와야 할 json key가 안들어오거나 Domain에 지정된 길이나 타입에 대한 제약을 위배하는 경우 모두 원하는 형식대로 처리는 안되고 있는 상황이라 함께 고민이 필요한 부분입니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/DTO/ResponseDTO.java:
```diff
@@ -0,0 +1,14 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ResponseDTO {
+    private String message;
+    private String errorCode;
```
에러코드를 전부 숫자로 이루어져 있는 형태로 설정했으므로 자료형도 정수를 사용하는 것이 좋겠습니다.
String을 사용하는 경우, 아무리 내부적으로 숫자만 사용하도록 정해져 있더라도
응답을 처리하는 쪽에서는 만에 하나 발생할 상황을 고려해 형변환 과정에서의 예외상황을 처리해야 합니다.

B2SIC replied:
정수형 타입으로 지정해서 수정하겠습니다.
그런데 에러 코드에서 사용되는 Method 번호나 오류 번호는 임의로 정하면 되는 건가요?
정수형으로 변경할 경우 Method 번호는 10 또는 11부터 시작 해야 할 것으로 보입니다.

namsic replied:
처음 제안에 의하면 method code는 각각 10과 11이지만 임의로 변경해도 무방합니다.

이와는 별개로 정수형 코드를 사용하는 경우 길이 기반 슬라이싱을 사용하지 않고
100으로 나눈 몫이나 나머지 값을 활용할 것이기 때문에 method code가 1부터 시작해도
문제 발생하지 않을 것으로 보입니다.

namsic reviewed (approved):
더 수정할 내용이 없으면 merge 하면 됩니다.

적용 이후 웹이나 앱에서 테스트할 수 있는 환경을 아래 방법 중 하나로 제공하는 것이 좋겠습니다.
선호하는 방법이나 다른 의견이 있다면 알려주기 바랍니다.
@regular1020 @chanw12
- 클라우드 활용
  - 장점: 실제 배포와 동일한 형태, 고정 ip 활용 가능
  다른 방법 적용해도 나중에는 클라우드 사용해야 함.
  - 단점: 비용 문제
- 개인 단말에서 git clone 및 직접 빌드
  - 장점: 변경사항 반영 가능, 로컬에서 테스트 가능
  - 단점: spring 개발을 하지 않음에도 개발환경을 구축해야 함
- 실행만 하면 동작하도록 빌드된 형태로 제공
  - 장점: 로컬에서 테스트 가능
  - 단점: OS 의존성이 있지는 않을지?

namsic commented:
`115.85.181.222:8080`에 5baf3cf0cefa9ae360d33ff26b7fe4a463ddb201 기준으로 구동해 두었습니다.

- @B2SIC
  - ~~요청은 http request에 json body로 보내는 것이 맞나요? 간단히 테스트해 보았는데 Bad Request 외 다른 응답 확인하지 못했습니다.~~
  추가: 알고 보니 find의 요청 format이 `{"phone_number": "123"}`이 아니라
  `{"type": "phone_number", "key": "123"}`이네요. wiki 등에 요청 및 응답 형식을 간단히 정리한 내용이 있으면 좋겠습니다.
  - IDE에서 직접 실행하는 것 외에 다른 방법으로 빌드 및 실행이 가능한지 확인 바랍니다. 서버에서 실행하기 위해 필요합니다.

- @regular1020 @chanw12
  - 동작 확인 및 의견 바랍니다.

참고로 현재 요청 형식은 아래와 같습니다.
`/member/join`
```json
{
  "user_id": "string",
  "password": "string",
  "phone_number": "string",
}
```
`/member/find`
```json
{
  "type": "user_id | phone_number",
  "key": "string"
}
```
